20220714-lookup-api.md

```diff
@@ -0,0 +1,460 @@
+              "user": "bob"
+            }
+          ]
+        }
+      }' \
+  localhost:8081 openfga.v1.OpenFGAService/Lookup
+
+{
+    "object_ids": [
+        "doc1",
+        "doc2",
+        "doc3",
+        "doc4
+    ]
+}
+```
+
+Sample curl command:
+```
+curl --request POST 'http://localhost:8080/stores/<storeID>/lookup' \
+--header 'Content-Type: application/json' \
+--data-raw '{
+    "authorization_model_id": "<modelID>"
+    "object_type": "document",
+    "relation": "viewer",
+    "user": "bob",
+    "contextual_tuples": {
+      "tuple_keys': [
+        {
+          "object": "document:doc4",
+          "relation": "viewer",
+          "user": "bob"
+        }
+      ]
+    }
+}'
+
+{
+    "object_ids": [
+        "doc1",
+        "doc2",
+        "doc3",
+        "doc4"
+    ]
+}
+```
+
+The results from the Lookup query can then be used to filter a collection by taking the intersection of these object ids with objects returned from your application's query.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The `StreamedLookup` endpoint will be (gRPC server stream) will be mapped to it's HTTP/json equivalent through the mapping provided by the implementation of the grpc-gateway project, which converts the stream into an [http/1.1 chunked transfer encoding](https://en.wikipedia.org/wiki/Chunked_transfer_encoding). Consumers of the streaming HTTP endpoint will be able to consume it using chunked-transfer encoding.
+
+The `Lookup` endpoint (unary gRPC endpoint) will be mapped to it's HTTP/json equivalent through the mapping provided by the implementation of the [grpc-gateway](https://grpc-ecosystem.github.io/grpc-gateway/) project.
+
+## API Changes
+The following is the proposed patch diff of the protobuf API changes to [openfga/api](https://github.com/openfga/api):
+```
+diff --git a/openfga/v1/openfga_service.proto b/openfga/v1/openfga_service.proto
+index 4140d56..d0056cb 100644
+--- a/openfga/v1/openfga_service.proto
++++ b/openfga/v1/openfga_service.proto
+@@ -661,6 +661,48 @@ service OpenFGAService {
+       description: "Returns a paginated list of OpenFGA stores."
+     };
+   }
++
++  // StreamedLookup is a streaming variation of the Lookup API (see Lookup for more info).
++  rpc StreamedLookup(LookupRequest) returns (stream StreamedLookupResponse) {
++    option (google.api.http) = {
++      post: "/stores/{store_id}/lookup:stream",
++      body: "*"
++    };
++  }
++
++  // Lookup looks up the object ids for objects of the provided type that the given user
++  // has a specific relation with.
++  rpc Lookup(LookupRequest) returns (LookupResponse) {
++    option (google.api.http) = {
++      post: "/stores/{store_id}/lookup",
++      body: "*"
++    }
++  }
++}
++
++message LookupRequest {
++    string store_id  = 1 [
++      json_name = "store_id"
++    ];
++
++    string authorization_model_id = 2 [
++      json_name = "authorization_model_id"
++    ];
++
++    string object_type = 3 [
```

Comment:
@rhamzeh had a comment about making this consistent with the other API definitions and simply calling it `type`. I think it would be nice to evolve our API definition to accept `object_type` and `object_id` instead of a single `object` string that we have to split etc.. It makes the API a little clearer as well, because 'type' is a very generic term.

I don't think this should be part of the scope of this work, but I thought I would mention it.